lib/model/recent_senders.dart

```diff
@@ -0,0 +1,154 @@
+      if (message is! StreamMessage) continue;
+      final StreamMessage(:streamId, :topic, :senderId, id: int messageId) = message;
+      (messagesByUserInStream[(streamId, senderId)] ??= []).add(messageId);
+      (messagesByUserInTopic[(streamId, topic, senderId)] ??= []).add(messageId);
+    }
+
+    for (final entry in messagesByUserInStream.entries) {
+      final (streamId, senderId) = entry.key;
+      ((streamSenders[streamId] ??= {})
+        [senderId] ??= MessageIdTracker()).addAll(entry.value);
+    }
+    for (final entry in messagesByUserInTopic.entries) {
+      final (streamId, topic, senderId) = entry.key;
+      (((topicSenders[streamId] ??= {})[topic] ??= {})
+        [senderId] ??= MessageIdTracker()).addAll(entry.value);
+    }
+  }
+
+  /// Records the necessary data from [message] if it is a [StreamMessage].
+  ///
+  /// If [message] is not a [StreamMessage], this is a no-op.
+  void handleMessage(Message message) {
+    if (message is! StreamMessage) return;
+    final StreamMessage(:streamId, :topic, :senderId, id: int messageId) = message;
+    ((streamSenders[streamId] ??= {})
+      [senderId] ??= MessageIdTracker()).add(messageId);
+    (((topicSenders[streamId] ??= {})[topic] ??= {})
+      [senderId] ??= MessageIdTracker()).add(messageId);
+  }
+
+  void handleDeleteMessageEvent(DeleteMessageEvent event, Map<int, Message> cachedMessages) {
+    if (event.messageType != MessageType.stream) return;
+    final messagesByUserInStream = <(int, int), List<int>>{};
+    final messagesByUserInTopic = <(int, String, int), List<int>>{};
+
+    final DeleteMessageEvent(:streamId!, :topic!) = event;
+    for (final id in event.messageIds) {
+      final message = cachedMessages[id] as StreamMessage?;
+      if (message == null) continue;
+      final senderId = message.senderId;
+      (messagesByUserInStream[(streamId, senderId)] ??= []).add(id);
+      (messagesByUserInTopic[(streamId, topic, senderId)] ??= []).add(id);
+    }
+
+    for (final entry in messagesByUserInStream.entries) {
+      final MapEntry(key: (streamId, senderId), value: messages) = entry;
+      final sendersMap = streamSenders[streamId];
+      final idTracker = sendersMap?[senderId];
+
+      idTracker?.removeAll(messages);
+      if (idTracker?.maxId == null) sendersMap?.remove(senderId);
+      if (sendersMap?.isEmpty ?? false) streamSenders.remove(streamId);
+    }
+
+    for (final entry in messagesByUserInTopic.entries) {
+      final MapEntry(key: (streamId, topic, senderId), value: messages) = entry;
+      final topicsMap = topicSenders[streamId];
+      final sendersMap = topicsMap?[topic];
+      final idTracker = sendersMap?[senderId];
+
+      idTracker?.removeAll(messages);
+      if (idTracker?.maxId == null) sendersMap?.remove(senderId);
+      if (sendersMap?.isEmpty ?? false) topicsMap?.remove(topic);
+      if (topicsMap?.isEmpty ?? false) topicSenders.remove(streamId);
+    }
+  }
+}
+
+class MessageIdTracker {
```

Comment:
Making this class private would cause the following lint message, thus failing the CI:
```
Invalid use of a private type in a public API.
Try making the private type public, or making the API that uses the private type also be private.
```